docs: Selfhosting on ubuntu from sources (#223)
cozy/cozy.github.io: src/tutorials/selfhosting/selfhost-ubuntu-sources-en.md

    In this docuemnt, we will use `domain.example` as an example. You will replace it with your own domain name throughout the explanation.
    The address of your cozy instance will be `cozy.domain.example`
- Good system administration knowledge. Despite documentation's goal is to be pretty straightforward to follow, there are some tricky and technical parts.

During installation, you will also define:

- a CouchDB admininstration password
- a CouchDB database access password
- a cozy-stack admin password
- You will need to provide your email address for Let's Encrypt SSL certificate validation and your cozy instance creation

# Couchdb

Configure CouchDB package repository

    sudo apt update && sudo apt install -y curl apt-transport-https gnupg
    curl https://couchdb.apache.org/repo/keys.asc | gpg --dearmor | sudo tee /usr/share/keyrings/couchdb-archive-keyring.gpg >/dev/null 2>&1
    echo "deb [signed-by=/usr/share/keyrings/couchdb-archive-keyring.gpg] https://apache.jfrog.io/artifactory/couchdb-deb/ $(lsb_release -sc) main" | sudo tee /etc/apt/sources.list.d/couchdb.list >/dev/null

> 

Reviewer: Does this syntax well supported by our mkdocs template? I know that sometimes our markdown is correctly displayed on github but not on docs.cozy.

Author: I'm happy with the visual result as tested on my laptop. or at least it's not worse than all our other documentation.
![image](https://user-images.githubusercontent.com/2768193/144563216-95f21762-5099-4b18-a7d7-8dfa1efa30d1.png)

Reviewer: Good to know thanks for taking the time to make the screenshot!

Install NodeJS

    sudo apt update
    sudo apt install -y nodejs
    sudo ln -s /usr/bin/node /usr/bin/nodejs

# Go

cozy-stack is developped in Go language so we need to install the go compiler to be able to compile cozy-stack sources.

    wget -O /tmp/go1.17.3.linux-amd64.tar.gz https://go.dev/dl/go1.17.3.linux-amd64.tar.gz
    sudo rm -rf /usr/local/go && sudo tar -C /usr/local -xzvf /tmp/go1.17.3.linux-amd64.tar.gz
    echo "export PATH=\"\$PATH:/usr/local/go/bin\"" | sudo tee /etc/profile.d/golang.sh > /dev/null
    source /etc/profile.d/golang.sh

Test Go installation is fine with

    go version
    go version go1.17.3 linux/amd64

# Cozy-stack

First, install requirements

    sudo apt install -y imagemagick libprotobuf-c1 fonts-lato

Get the source code

    sudo apt install -y git
    sudo git clone https://github.com/cozy/cozy-stack.git /opt/cozy-stack

Then compile the program

    cd /opt/cozy-stack
    scripts/build.sh release $(go env GOPATH)/bin/cozy-stack

The compilation generate a binary file under `$GOPATH/bin/cozy-stack`

You can test it with :

    $(go env GOPATH)/bin/cozy-stack version

>     1.5.0-5-gcbdf012d

Reviewer: nit: Since we install from https://go.dev/dl/go1.17.3.linux-amd64.tar.gz, the version should be 1.17.3
```suggestion
   1.17.3
```

Finally, configure systemd to automatically launch cozy-stack on boot.

    cat <<EOF | sudo tee /usr/lib/systemd/system/cozy-stack.service >/dev/null
    [Unit]
    Description=Cozy service
    Wants=couchdb.service
    After=network.target couchdb.service

    [Service]
    User=cozy-stack
    Group=cozy
    WorkingDirectory=/var/lib/cozy/
    PermissionsStartOnly=true
    ExecStart=/usr/bin/cozy-stack serve
    Restart=always

    [Install]
    WantedBy=multi-user.target
    EOF
    sudo systemctl daemon-reload
    sudo systemctl enable cozy-stack
    sudo systemctl start cozy-stack

You can validate everything went well and cozy-stack is running thiw way

    curl http://localhost:8080/version
    {"build_mode":"production","build_time":"2021-12-01T13:12:36Z","runtime_version":"go1.17.3","version":"1.5.0-5-gcbdf012d"}

# Nginx

First create a DNS entry in your domain for `cozy.domain.example` and `*.cozy.domain.example` pointing at your server. For example :

    cozy     1h     IN         A     <your_server_IP>
    *.cozy   1h     IN     CNAME     cozy

Then install Nginx

    sudo apt install -y nginx certbot

Generate SSL certificate with certbot

    DOMAIN=domain.example

>     EMAIL="<votre email>"

Reviewer: ```suggestion
    EMAIL="<your mail address>"
```

    sudo certbot certonly --email "${EMAIL}" --non-interactive --agree-tos --webroot -w /var/www/html -d cozy.${DOMAIN} $(printf -- " -d %s.cozy.${DOMAIN}" home banks contacts drive notes passwords photos settings store)

> Create nginx reload script for your certificate to be reloaded each time it is automatically refreshed, every 3 monthes

Reviewer: ```suggestion
Create nginx reload script for your certificate to be reloaded each time it is automatically refreshed, every 3 months
```

        ssl_certificate_key /etc/letsencrypt/live/cozy.${DOMAIN}/privkey.pem;

        server_name *.cozy.${DOMAIN} cozy.${DOMAIN};
        access_log /var/log/nginx/cozy.${DOMAIN}.access.log;
        error_log /var/log/nginx/cozy.${DOMAIN}.error.log;

        add_header Strict-Transport-Security "max-age=31536000; includeSubDomains;";
        client_max_body_size 1g;

        location / {
            proxy_pass http://localhost:8080;
            proxy_http_version 1.1;
            proxy_set_header Upgrade \$http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_set_header Host \$host;
            proxy_set_header X-Forwarded-For \$remote_addr;
        }
    }
    EOF
    sudo ln -s ../sites-available/cozy.${DOMAIN} /etc/nginx/sites-enabled/
    sudo systemctl reload nginx

You can then test from your browser by visiting `https://cozy.domain.example` and you should see a page telling you this cozy instance doesn't exist yet. This is the sign that everything went well and the only part left is to create the instance.

# Cozy instance creation

Create your cozy instance

    DOMAIN=domain.example

>     EMAIL=<votre email>

Reviewer: ```suggestion
    EMAIL=<your mail address>
```

    [[ -z "${COZY_PASS}" ]] && read -p "Cozy stack admin password: " -r -s COZY_PASS

>     COZY_ADMIN_PASSWORD="${COZY_PASS}" cozy-stack instances add --apps home,banks,contacts,drive,notes,passwords,photos,settings,store --email "${EMAIL}" --locale fr --tz "Europe/Paris" cozy.${DOMAIN}

Reviewer: ```suggestion
    COZY_ADMIN_PASSWORD="${COZY_PASS}" cozy-stack instances add --apps home,banks,contacts,drive,notes,passwords,photos,settings,store --email "${EMAIL}" --locale fr --en "Europe/Paris" cozy.${DOMAIN}
```

Author: I think you mean `--locale en` instead of `--en`

Beacause our main audience is currently in france, I prefer to leave the french locale and Paris timezone in the command example.

Anyway, the text below explain how to change that.

Reviewer: Oh sh** yes this is what I meant 😵‍💫 . I saw just after the next line but kept the comment to see what do you though about it

You can of course adapt your language (`locale`) and choose english (`en`) or spanish (`es`) and choose another timezone (`tz`).

Note the “Registration token” this command returns and visit from your browser `https://cozy.domain.example?registerToken=<registration_token>` substituting `domain.example` with your real domain name and `<registration_token>` with the “Registration token” you got.

> You will be promptes to define your cozy password and you will be able to start using your self-hosted cozy.

Reviewer: ```suggestion
You will be prompted to define your cozy password and you will be able to start using your self-hosted cozy.
```

# Et voilà !

Your cozy is not fully operational! Its address is `https://cozy.domain.example` (remplace `domain.example` by your own domain name)

> You can then start installing connectors from store to automatically gather your data, sauve your passwords in cozy-pass, store your filesin cozy-drive and install cozy-desktop client on your PC to synchronize your cozy content with a local folder.

Reviewer: ```suggestion
You can then start installing connectors from store to automatically gather your data, save your passwords in cozy-pass, store your files in cozy-drive and install cozy-desktop client on your PC to synchronize your cozy content with a local folder.
```

Below are some bonuses 😉

# Hosting more than one cozy instance on the same server

Having its own selfhosted cozy instance is nice but hosting cozy instances for friends and familly is a must! Here is how to add more cozy instances on the same server.

The first cozy instance we added was `https://cozy.domain.example`. We will create a second cozy instance for Mary with address `https://mary.domain.example` (Replace `domain.example` with your own domain name and `mary` whith what you want to uniquely identify the cozy instance.

So we will need :

- Our domain name. We still use `domain.example` in this documentation
- The new cozy instance's “slug”, which is its unique identifier. We will use `mary` here for example. The address for this new cozy instance will the be in the form `https://<slug>.<domain>`, for example here `https://mary.domain.example`

First, let's put all that important information in variables

    DOMAIN=domain.example

>     EMAIL=<votre adresse email>

Reviewer: ```suggestion
    EMAIL=<your mail address>
```

    NEWSLUG=mary

>     NEWEMAIL=<adresse email de mary>

Reviewer: ```suggestion
    NEWEMAIL=<mary's mail address>
```

        ssl_certificate /etc/letsencrypt/live/${NEWSLUG}.${DOMAIN}/fullchain.pem;
        ssl_certificate_key /etc/letsencrypt/live/${NEWSLUG}.${DOMAIN}/privkey.pem;

        server_name *.${NEWSLUG}.${DOMAIN} ${NEWSLUG}.${DOMAIN};
        access_log /var/log/nginx/${NEWSLUG}.${DOMAIN}.access.log;
        error_log /var/log/nginx/${NEWSLUG}.${DOMAIN}.error.log;

        add_header Strict-Transport-Security "max-age=31536000; includeSubDomains;";
        client_max_body_size 1g;

        location / {
            proxy_pass http://localhost:8080;
            proxy_http_version 1.1;
            proxy_set_header Upgrade \$http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_set_header Host \$host;
            proxy_set_header X-Forwarded-For \$remote_addr;
        }
    }
    EOF
    sudo systemctl reload nginx

Create cozy instance

    [[ -z "${COZY_PASS}" ]] && read -p "Cozy stack admin password: " -r -s COZY_PASS
    COZY_ADMIN_PASSWORD="${COZY_PASS}" cozy-stack instances add --apps home,banks,contacts,drive,notes,passwords,photos,settings,store --email "${NEWEMAIL}" --locale fr --tz "Europe/Paris" ${NEWSLUG}.${DOMAIN}

> Note the “Registration token” the last command gives you and send Mary the following url: `https://mary.domain.example?registerToken=<registration_token>`, substituting `domain.example` with your own domain name, `mary` with the slug you chose for this new instance and  `<registration_token>` with the “Registration token” rturned by the last command.

Reviewer: ```suggestion
Note the “Registration token” the last command gives you and send Mary the following url: `https://mary.domain.example?registerToken=<registration_token>`, substituting `domain.example` with your own domain name, `mary` with the slug you chose for this new instance and  `<registration_token>` with the “Registration token” returned by the last command.
```

    sudo -i -u postgres psql -c "GRANT ALL privileges ON DATABASE onlyoffice TO onlyoffice;

The second command create a database user names `onlyoffice` with password `onlyoffice`. We advise you to choose a more secure password in real life.

### Install RabbitMQ

    sudo apt install -y rabbitmq-server

### Install Onlyoffice Documentserver

Configure package repository

    sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys CB2DE8E5
    echo "deb https://download.onlyoffice.com/repo/debian squeeze main" | sudo tee /etc/apt/sources.list.d/onlyoffice.list
    sudo apt update

Install microsoft fonts

    sudo apt install -y ttf-mscorefonts-installer

When asked, accept EULA

Install OnlyOffice Documentserver

    sudo apt install -y onlyoffice-documentserver

When asked, enter database password we created when installing postgreSQL and created database.

Then restart Nginx

    sudo systemctl reload nginx

### Configure HTTPS for OnlyOffice

Create a DNS entry for OnlyOffice targetting your server. For example :

    onlyoffice     1h     IN         A     <your_server_IP>

Generate SSL certificate

>     EMAIL=<votre email>

Reviewer: ```suggestion
    EMAIL=<your mail address>
```
